File: certificate-manager/src/types/certificate.ts
/**
 * Certificate type matching @certificate-service backend API.
 * GET /certificates/:id returns this shape.
 * Backend: domain/certificate.rs (id, subject, issuer, valid_from, expiration, signature_algorithm, san_entries).
 */
export interface Certificate {
  id: string;
  subject: string;
  issuer: string;
  /** Validity start (notBefore) ISO 8601 UTC. Present when parsed from PEM; null when loaded from list without re-parsing. */
  valid_from: string | null;
  /** ISO 8601 UTC datetime (notAfter, e.g. from chrono::DateTime<Utc>). */
  expiration: string;
  /** Signature algorithm (e.g. sha1WithRSAEncryption). Present when parsed from PEM; null otherwise. */
  signature_algorithm: string | null;
  /** Subject Alternative Names (DNS, IP, etc.). */
  san_entries: string[];
  /** Last time the record was updated (ISO 8601). */
  last_updated?: string | null;
}

/**
 * Generic paginated list response for GET /certificates?page=&limit=
 * Use when the backend implements a list endpoint.
 */
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

/**
 * Dashboard metrics (e.g. from GET /certificates/metrics or derived from list).
 * Backend may expose these when list/stats endpoints are added.
 */
export interface DashboardMetrics {
  total: number;
  expiring_soon: number; // e.g. next 30 days
}

/**
 * Certificate status derived from expiration (client-side).
 */
export type CertificateStatus = "Valid" | "Expiring Soon" | "Expired";

/**
 * Helper: derive status from expiration ISO string.
 */
export function getCertificateStatus(expirationIso: string): CertificateStatus {
  const exp = new Date(expirationIso);
  const now = new Date();
  const thirtyDaysFromNow = new Date(now);
  thirtyDaysFromNow.setDate(thirtyDaysFromNow.getDate() + 30);
  if (exp < now) return "Expired";
  if (exp <= thirtyDaysFromNow) return "Expiring Soon";
  return "Valid";
}

File: certificate-service/src/domain/certificate.rs
//! Domain entity: Certificate.
//! No dependencies on SQLx, Axum, or x509-parser.
//! `expiration` and `valid_from` serialize as ISO 8601 (RFC 3339) via chrono's serde support.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Certificate {
    pub id: String,
    pub subject: String,
    pub issuer: String,
    /// Validity start (notBefore) in UTC. Present when parsed from PEM; null when loaded from DB without re-parsing.
    pub valid_from: Option<DateTime<Utc>>,
    /// Expiration time (not_after) in UTC. Serializes as ISO 8601 string.
    pub expiration: DateTime<Utc>,
    /// Signature algorithm (e.g. sha1WithRSAEncryption). Present when parsed from PEM; null when loaded from DB without re-parsing.
    pub signature_algorithm: Option<String>,
    /// Subject Alternative Names (DNS, IP, etc.).
    pub san_entries: Vec<String>,
    /// Last time the record was updated (for display in inventory).
    pub last_updated: Option<DateTime<Utc>>,
}

impl Certificate {
    pub fn new(
        id: String,
        subject: String,
        issuer: String,
        valid_from: Option<DateTime<Utc>>,
        expiration: DateTime<Utc>,
        signature_algorithm: Option<String>,
        san_entries: Vec<String>,
        last_updated: Option<DateTime<Utc>>,
    ) -> Self {
        Self {
            id,
            subject,
            issuer,
            valid_from,
            expiration,
            signature_algorithm,
            san_entries,
            last_updated,
        }
    }
}
